pkg/gatewayconfigurationserver/gatewayconfigurationserver.go

```diff
@@ -0,0 +1,155 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayconfigurationserver
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gogo/protobuf/types"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/basicstation/cups"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	ttnweb "go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc/metadata"
+)
+
+// Config contains the Gateway Configuration Server configuration.
+type Config struct {
+	// BasicStation defines the configuration for the BasicStation CUPS server.
+	BasicStation cups.ServerConfig `name:"basic-station" description:"BasicStation CUPS server configuration."`
+	// RequreAuth defines if the HTTP endpoints should require authentication or not.
+	RequireAuth bool `name:"require-auth" description:"Require authentication for the HTTP endpoints."`
+}
+
+// GatewayConfigurationServer implements the Gateway Configuration Server component.
+type GatewayConfigurationServer struct {
+	*component.Component
+	config *Config
+
+	registry ttnpb.GatewayRegistryClient
+
+	ctx context.Context
+}
+
+// RegisterRoutes registers the web frontend routes.
+func (gcs *GatewayConfigurationServer) RegisterRoutes(server *ttnweb.Server) {
+	middleware := []echo.MiddlewareFunc{
+		gcs.handleError(),
+		gcs.validateAndFillIDs(),
+	}
+	if gcs.config.RequireAuth {
+		middleware = append(middleware, gcs.requireGatewayRights(ttnpb.RIGHT_GATEWAY_INFO))
+	}
+	group := server.Group(ttnpb.HTTPAPIPrefix+"/gcs/gateways/:gateway_id", middleware...)
+	group.GET("/global_conf.json", func(c echo.Context) error {
+		return gcs.handleGet(c)
+	})
+}
+
+// New returns new *GatewayConfigurationServer.
+func New(c *component.Component, conf *Config, opts ...Option) (*GatewayConfigurationServer, error) {
+	gcs := &GatewayConfigurationServer{
+		Component: c,
+		config:    conf,
+		ctx:       c.FillContext(c.Context()),
+	}
+	for _, opt := range opts {
+		opt(gcs)
+	}
+
+	cups := conf.BasicStation.NewServer(c)
+	_ = cups
+
+	c.RegisterWeb(gcs)
+	return gcs, nil
+}
+
+// Option represents a Gateway Configuration Server option.
+type Option func(gcs *GatewayConfigurationServer)
+
+// WithRegistry sets the gateway registry for the given server.
+func WithRegistry(registry ttnpb.GatewayRegistryClient) Option {
+	return func(gcs *GatewayConfigurationServer) {
+		gcs.registry = registry
+	}
+}
+
+// WithContext sets the context for the given server.
+func WithContext(ctx context.Context) Option {
```

Comment:
Why would we replace the context? We should inherit it from the Component, right?

Reply:
I'm using it inside the tests - It is required for the `rights` fetcher.

Reply:
We shouldn't be doing that. If you need a rights fetcher, consider spinning up a mock IS, or make sure a fetcher is available in the context where you need it.

Reply:
This is not addressed